### ISSUE TITLE
Allow handling of null for Convert of BaseConverterOneWay<TFrom, TTo>

### ISSUE THREAD

**Issue by Countryen** [CommunityToolkit/Maui#184]
### Discussed in https://github.com/CommunityToolkit/Maui/discussions/164

<div type='discussions-op-text'>

<sup>Originally posted by **Countryen** October 22, 2021</sup>
The base class for IValueConverter `BaseConverterOneWay<TFrom, TTo>` provides many features/benefits but doesn't allow overriding or defining what happens when a `null` is passed into the `Convert` method.

It would be very nice if it was able to either override or define the type/value checking or to set a "allows null" for the value.

This helps using converters with strings that can also be `null`.
Is there a better solution?</div>

Most important details from the discussion:
* The `TargetNullValue` of XAML's `Converter` for bindings is applied to the result of the converter and not to the value before it goes in to the converter.
* Let the BaseConverterOneWay<TFrom, TTo> check for null and fallback to the default value for TFrom instead of throwing an Exception. (Suggested by @pictos )
* > If we went with a blanket always replace nulls with default then that is a behavioral change but is still probably unlikely to have a big impact. If we added a way to opt-in as @Countryen suggested then it would have even less of an impact.
* This would then mean that the actual custom converter's ConvertFrom(TFrom value) is called with default(TFrom) when value is null. We could then add a null check to return something the view can use.

**Comment by bijington** [comment]
@Countryen thank you for raising this issue. Given that you have an understanding is this something you feel you could contribute to fixing? There is of course no pressure though 😄

**Comment by Countryen** [comment]
@bijington Would love to contribute. What do I have to do? 
Is it enough to create a PR and wait for automatic test runs or do I have to run tests locally on my device first? Do I have to test MAUI and Xamarin Forms or only Xamarin Forms? I can't use MAUI in the office, currently. Do you expect a working sample or unit test to be added?

**Comment by bijington** [comment]
@Countryen that is great thank you. To answer your questions:

> Is it enough to create a PR and wait for automatic test runs or do I have to run tests locally on my device first?

Ideally we would like as much tested as possible. It is possible to run the unit tests locally if that helps and if changing behavior as we would be with this change additional unit tests would be great. Obviously not everyone has access to all device types (Android, iOS, etc.) so if you are not able to test some devices then just letting us know in the PR would be really helpful. That said this change is extremely unlikely to have platform specific issues.

> Do I have to test MAUI and Xamarin Forms or only Xamarin Forms?

So this is where it may get a little bit more confusing but this issue is raised against the MAUI Community Toolkit and is completely separate to the [Xamarin Community Toolkit](https://github.com/xamarin/XamarinCommunityToolkit). We will need to fix this in both toolkits to retain consistency but there is no expectation to provide both, just as much as you feel you can. If you are in a better position to test/fix for the Xamarin Community Toolkit then it would probably best to fix [this issue](https://github.com/xamarin/XamarinCommunityToolkit/issues/1705). I just raised a copy of this issue against the Xamarin Community Toolkit to cover the change there.

> Do you expect a working sample or unit test to be added?

Yes please. Given that you would be fixing something and not adding new functionality it should hopefully be possible to just tweak an existing page/unit test class to expose this new behavior rather than having to create something new. We are hear to answer any questions though if anything isn't clear in what you would need to change 😄 


In summary I think it would probably be best to look at the XCT issue first. What are your thoughts?

And of course please feel free to ask for guidance on any of these steps.

**Comment by Countryen** [comment]
@bijington I cloned the XCT repo and ran the unit tests, thanks for the info above. I've noticed that:

Some unit tests are already red for me in the DateTimeOffsetConverter_Tests. This is probably because I use a German system.
If wanted I can report this in a separate issue.

---

Also, I've noticed some small spelling errors (in the tests). Should I only fix the topic of this issue and ignore them?
For example: BoolToObjectInValidValuesThrowArgumen**t**Exception
(I think it should be BoolToObject**Invalid**ValuesThrow**s**Argumen**t**Exception, but only the missing **t** in Argumen**t** is what I see important).

---

Sadly, there is no test that uses the `BaseConverterOneWay<TFrom, TTo>`. The only classes that derive from this are the **color conversion classes** but they are not tested at all it seems.

So I will have to add a new unit test class like (ColorToRgbStringConverter_Tests) and add a test for the invalid arguments (like done in other test classes) where `null` should be possible.

If not wanted I can also just add a custom private implementation in the test class itself to test it.

**Comment by bijington** [comment]
@Countryen that's great you have been able to start on this. To answer your questions:

> Some unit tests are already red for me in the DateTimeOffsetConverter_Tests. This is probably because I use a German system.
If wanted I can report this in a separate issue.

If you could raise an issue that would be fantastic. Sadly we do have some inconsistent tests that we need to tidy up and I might see if I can get those sorted to reduce any friction for new contributors.

> Also, I've noticed some small spelling errors (in the tests). Should I only fix the topic of this issue and ignore them?
For example: BoolToObjectInValidValuesThrowArgumentException
(I think it should be BoolToObjectInvalidValuesThrowsArgumentException, but only the missing t in Argument is what I see important).

Ideally if we can keep a PR limited to the scope of the issue/feature that is ideal as it makes it easier to review. If the tests are in the same area you are working though I see no reason why you couldn't/shouldn't fix typos like this.

> Sadly, there is no test that uses the BaseConverterOneWay<TFrom, TTo>. The only classes that derive from this are the color conversion classes but they are not tested at all it seems.

> So I will have to add a new unit test class like (ColorToRgbStringConverter_Tests) and add a test for the invalid arguments (like done in other test classes) where null should be possible.

> If not wanted I can also just add a custom private implementation in the test class itself to test it.

I think adding a test like you suggest `ColorToRgbStringConverter_Tests` sounds like a good approach as it not only tests your new code but provides the ability to test pre-existing code that should have already been tested.

Thanks again for looking at this

**Comment by Countryen** [comment]
Happy new year. After resolving this last year in XCT it would now make sense to start changing this for MCT.
What is the process? Is the change from XCT (automatically) also done for MCT or do we have to manually copy-paste most of the changes/tests?

We/You did speak about a different approach because of added possibilities (like "is null if" for nullability).
I think it's important to have this feature in MAUI too, otherwise migrating might be weird, if someone uses the new allowed null handling already.

Is it okay if we just copy-paste the old solution for from XCT for now and create a new issue that "improves" the solution or do you want to give it more thought from the get go?

I am happy to assist as much as I can - I have no experience with MAUI or the MCT, though.

**Comment by brminnick** [comment]
Yes! Let's [open a Discussion in the .NET MAUI Community Toolkit repo](https://github.com/CommunityToolkit/Maui/discussions/new) to chat about the best way to implement it.

Like @bijington mentioned, our implementation in `CommunityToolkit.Maui` won't be an exact copy/past because it is structured differently than `Xamarin.CommunityToolkit`.

For example, `CommunityToolkit.Maui.Converters.BaseConverterOneWay` doesn't currently throw an exception for `null`:

https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui/Converters/BaseConverterOneWay.shared.cs

```cs
using System.Globalization;
using CommunityToolkit.Maui.Extensions;

namespace CommunityToolkit.Maui.Converters;

/// <summary>
/// Abstract class used to implement converters that implements the ConvertBack logic.
/// </summary>
/// <typeparam name="TFrom">Type of the input value</typeparam>
/// <typeparam name="TTo">Type of the output value</typeparam>
public abstract class BaseConverterOneWay<TFrom, TTo> : ValueConverterExtension, ICommunityToolkitValueConverter
{
	/// <summary>
	/// Converts the incoming value from <typeparamref name="TFrom"/>[] and returns the object of a type <typeparamref name="TTo"/>.
	/// </summary>
	/// <param name="value">The value to convert.</param>
	/// <param name="targetType">The type of the binding target property. This is not implemented.</param>
	/// <param name="parameter">Additional parameter for the converter to handle. This is not implemented.</param>
	/// <param name="culture">The culture to use in the converter. This is not implemented.</param>
	/// <returns>An object of type <typeparamref name="TTo"/>.</returns>
	public object? Convert(object? value, Type? targetType, object? parameter, CultureInfo? culture)
	{
		if (value is not TFrom valueFrom)
			throw new ArgumentException($"value needs to be of type {typeof(TFrom)}");

		if (targetType != typeof(TTo) && !(typeof(TFrom) != typeof(string)))
			throw new ArgumentException($"targetType needs to be typeof {typeof(TTo)}");

		return ConvertFrom(valueFrom);
	}

	/// <summary>
	/// Method that will be called by <see cref="Convert(object, Type, object, CultureInfo)"/>.
	/// </summary>
	/// <param name="value">Value to be converted from <typeparamref name="TFrom"/> to <typeparamref name="TTo"/>.</param>
	/// <returns>An object of type <typeparamref name="TTo"/>.</returns>
	public abstract TTo? ConvertFrom(TFrom value);

	/// <summary>
	/// Not implemented, use <see cref="BaseConverter{TFrom, TTo}"/>
	/// </summary>
	public virtual object? ConvertBack(object? value, Type? targetType, object? parameter, CultureInfo? culture)
		=> throw new NotImplementedException("Impossible to revert to original value. Consider setting BindingMode to OneWay.");
}
```

Closing this Issue per [#1733](https://github.com/xamarin/XamarinCommunityToolkit/pull/1733)

**Comment by Countryen** [comment]
@bijington Thanks for the information. Interesting choice with the "is null", I am excited to see the changes.
Of course, best to take all the time needed to implement the feature most appropriately for MCT.

I will let you guys handle that for now as I don't follow MAUI / MCT development as much as XCT, yet. I will keep an eye on it and maybe add tests again or something to support later.
